iceoryx2-ffi/cxx/include/iox2/waitset.hpp

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#ifndef IOX2_WAITSET_HPP
+#define IOX2_WAITSET_HPP
+
+#include "iox/duration.hpp"
+#include "iox/expected.hpp"
+#include "iox2/file_descriptor.hpp"
+#include "iox2/internal/iceoryx2.hpp"
+#include "iox2/listener.hpp"
+#include "iox2/service_type.hpp"
+#include "iox2/waitset_enums.hpp"
+
+namespace iox2 {
+/// The [`Guard`] is returned by [`WaitSet::attach_deadline()`], [`WaitSet::attach_notification()`]
+/// or [`WaitSet::attach_interval()`]. As soon as it goes out-of-scope it detaches the attachment.
+/// It can also be used to determine the origin of an event in [`WaitSet::run()`] or
+/// [`WaitSet::run_once()`] via [`AttachmentId::event_from()`] or [`AttachmentId::deadline_from()`].
+template <ServiceType S>
+class Guard {
```

Comment:
Is this name to generic ? Wondering if other types of guards will be needed within the `iox2` namespace. This could be changed when the need for another type of `Guard` comes though.